Definitions for Device, Service, etc.

mmccool opened the pull request:
New definitions for Device, Service, and Shadow.
Updated definitions for Virtual Thing and Digital Twin.
Resolves issue #709.
Resolves issue #708.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/mmccool/wot-architecture/pull/711.html" title="Last updated on Feb 17, 2022, 5:15 PM UTC (14d0d30)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-architecture/711/78521c3...mmccool:14d0d30.html" title="Last updated on Feb 17, 2022, 5:15 PM UTC (14d0d30)">Diff</a>

benfrancis commented on index.html:
```diff
@@ -716,8 +748,10 @@ <h1>Terminology</h1>
-      <dd>An instance of a Thing that represents a Thing that is located
-        on another system component.</dd>
+      <dd>A <a>Service</a> that represents another <a>Thing</a>.
+       A Virtual Thing will often as as an <a>Intermediary</a>.
```
typo: "as as"

benfrancis commented on index.html:
```diff
@@ -444,15 +444,28 @@ <h1>Terminology</h1>
+      <dt>
+        <dfn>Device</dfn>
+      </dt>
+      <dd>A Device is a physical entity that has a network interface.
+      Devices can be described by a <a>Thing Description</a> and are a kind of <a>Thing</a>.
```
This is really a "Connected Device", but "Device" is a useful shorthand. I suggest adding both terms to this definition since both terms are used in the body of the text.

mlagally replied:
Arch call on 17.2.: Web Device?
Review usages of connected device?

mlagally replied:
I agree with @benfrancis , let's add Connected Device. We could also include Network Device as a synonym.

benfrancis replied:
Just to note that since the terms "Thing" and "Web Thing" have been used as synonyms up until this point [1], giving the two terms separate definitions may create confusion and would require a full review of all specifications to make sure the terms are being used correctly.

Does it really make sense to say that a Thing that isn't a Web Thing can be described by a Web of Things Thing Description?

I won't rehash my arguments for why "MQTT Web Thing" makes no sense ;)

> I suppose you can say "HTTP Web Thing" if necessary (and maybe this is a good name for the Profile we are working on...).

I wouldn't object to that :)

--- 

I have personally often used "Thing" to describe the physical device in the real world represented on the web by a "Web Thing", but that is probably not the common understanding. Those concepts would probably be described as "Device" and "Web Thing" respectively in this latest terminology.

mmccool replied:
Anyways, as I said elsewhere, that is also why I hesitate to put "Web" in front of "Service", since it may imply use of HTTP for some people.  So for now I will also leave that alone.

mmccool replied:
We should review how Web Thing is used.  It may be some cleanup is needed to map (Web Thing, Thing) to (Thing, Device).

mmccool replied:
I added "Connected Device" but did not change the definition of Web Thing.  See discussion below.

benfrancis commented on index.html:
```diff
@@ -716,8 +748,10 @@ <h1>Terminology</h1>
-        on another system component.</dd>
+      <dd>A <a>Service</a> that represents another <a>Thing</a>.
+       A Virtual Thing will often as as an <a>Intermediary</a>.
+       Examples include <a>Shadows</a> and <a>Digital Twins</a>.
```
I too have previously used the term "Virtual Thing" to mean a Web Thing which has no physical device associated with it. E.g. WebThings Gateway has a "Virtual Things add-on" which lets you add all kinds of virtual devices to your gateway which don't exist in the real world. This is useful for testing, but also for use cases like creating a virtual on/off switch which controls multiple physical devices for example.

What would be a good term for describing a Web Thing which has no physical counterpart? Could this be another kind of Virtual Thing?

mmccool replied:
hmm... you also use "virtual thing" for a "testing service" that "stands in place of" a physical thing.  I guess I could add that to the definition.  In this case, there is still "another thing" that the virtual thing is "standing in for".

benfrancis replied:
> In this case, there is still "another thing" that the virtual thing is "standing in for".

Not necessarily, it could be a virtual instance of a device which doesn't have a physical counterpart. As I said...

>> but also for use cases like creating a virtual on/off switch which controls multiple physical devices for example.

It still has all the properties of a Web Thing (a Thing Description, interaction affordances with network interfaces), it just doesn't represent an actual physical object in the real world. As far as a Consumer of a Web Thing is concerned the distinction is irrelevant, it doesn't know or care whether changes to the Web Thing manifest in the physical world.

mmccool replied:
I meant by "standing in for" that the "virtual instance" was standing in for a "real instance".  Anyhow, please take a look at the updated definition.

benfrancis reviewed (approved):
A couple of small nits, but this is really helpful. I found the definitions of Shadow and Digital Twin to be particularly helpful because I've always struggled with the difference between a Web Thing and a Digital Twin.

I've been reluctant so far to accept the idea that any "service" can be a "thing" because it may create the perception that the Thing Description can and should be used to describe any kind of web service (like OpenAPI), and that could lead to significant scope creep. However, I think the `DirectoryThing` in WoT Discovery makes it hard to continue to argue that point since that is clearly a web service, which isn't a physical device, being described by a Thing Description.

Thumbs up from me 👍

mlagally commented on index.html:
```diff
@@ -631,13 +644,32 @@ <h1>Terminology</h1>
+      </dt>
+      <dd>A Shadow is a <a>Virtual Thing</a> that
+      maintains a copy of the state and mediates interactions with another
+      <a>Thing</a> which may not be immediately available due to network
```
I suggest to define a shadow without statements about whether the thing is available or not.
IMHO a shadow reflects the state of another thing.

mlagally commented on index.html:
```diff
@@ -716,8 +748,10 @@ <h1>Terminology</h1>
       </dt>
-      <dd>An instance of a Thing that represents a Thing that is located
-        on another system component.</dd>
+      <dd>A <a>Service</a> that represents another <a>Thing</a>.
```
or thing*s*?

mmccool replied:
See above comments to Ben.    I plan to broaden the definition a bit/a lot and include this.  But this does not mean it's only a collection: a Virtual Thing is still a Service, so it's still a software entity with a network interface.  But yeah, there are use cases where we want to use a Virtual Thing to interface to multiple other Things.

My main concern right now is that my proposed extended definition of Virtual Thing is sounding more and more like "Intermediary" and I want to avoid having two terms for the same idea.  I will leave it alone for now, but options are: (1) restricting Intermediaries to one Exposer (2) making Virtual Things a kind of Intermediary, but also allowing Devices (e.g. physical Things) to be Intermediaries.  I'm leaning towards the second option.  I think it would be odd to call a physical device a "virtual" Thing.

mlagally commented on index.html:
```diff
@@ -631,13 +644,32 @@ <h1>Terminology</h1>
+      connectivity, power savings modes, or other factors.
+      A Shadow aims to achive eventual consistency with the state
+      of the Thing it represents.
+      If a Shadow has more advanced functionality than simply caching state
```
What do you mean by "advanced functionality"? Could this be interactions via other (= non-property) affordances? 
Is a event consumer still a shadow or is it already a twin?

I believe there's no clear border between shadows and twins, suggest to use these terms as synonyms.

benfrancis replied:
I also wasn't sure about the distinction between a Shadow and a Digital Twin either, but the distinction being one-way (physical device to shadow) vs. two-way (between physical device and twin) makes a lot of sense. I haven't done a comprehensive literature review but this does seem to match the definition used by organisations like [Fraunhofer IESE](https://www.iese.fraunhofer.de/en/services/digital-twin.html), [MDPI](https://www.mdpi.com/2075-5309/11/4/151/pdf) and [PSA](https://www.psa.inc/company/news/digital-model-vs-digital-shadow-vs-digital-twin-and-their-place-in-eiot/).

mmccool replied:
I would prefer to have a Shadow JUST mirror state, and be specific about that.  It's a lot easier to reason about Shadows as a part of a system if that's all it does, e.g. for the case of a hub having to deal with client devices that need to sleep and just call in occasionally on their own schedule to sync up.  If it's a Shadow we can reason about it like a slow version of the actual device.  If it's a Digital Twin it could be doing anything.
 
We could make Shadow a special case of a Digital Twin but then it's unclear whether we JUST mean mirroring when we say Digital Twin.  I would prefer to reserve Digital Twin for cases where it does more than just mirroring.

mmccool replied:
One more minor technical point: hubs may have Things that "represent" other devices that may not actually have Thing Descriptions, for example a Service in a hub may present a network interface and shadow some other client device without a Thing Description, like a Z Wave or Zigbee device.  Not sure how to deal with that right now, we may have to loosen the definitions slightly to allow for shadowing of entities that don't have actual Thing Descriptions and/or that Thing Descriptions can't currently handle.

mmccool replied:
Re one-way vs two-way: I will have to look at the definitions you cite e.g. by Fraunhofer.  But to me the sync would go both ways.  Suppose I have a switch that for some weird reason is only a client and only syncs periodically.  The I want to "read" a binary property on the Shadow to know what state the switch is in, but I might also want to "write" the property to flip the state.  In this case the real switch (when it calls in) should check the state on the switch and sync with it.   So then the Shadow acts like a "slow" version of the real device, with a latency (at most) equal the interval at which the client device wakes up and checks in.

Anyway, I think they are different, if not for precisely this reason.

mlagally commented on index.html:
```diff
@@ -631,13 +644,32 @@ <h1>Terminology</h1>
       <dd>The combination of Public Security Metadata, Private Security Data, and any other configuration
         information (such as public keys) necessary to operationally configure the security mechanisms of a Thing.</dd>
+      <dt>
+        <dfn>Service</dfn>
```
Should we be using the term WebService?

mmccool replied:
In particular, if it is a software component without a network interface, it is usually called a "module" or a "library".  When I hear the word "service" I immediately assume it has a network interface.

mmccool commented on index.html:
```diff
@@ -4270,10 +4304,13 @@ <h1>Recent Specification Changes</h1>
       <li>New chapter: Lifecycle.</li>
       <li>New terminology: Thing Model.</li>
-      <li>Chapter restructuring and renaming:
+      <li>Chapter restructuring and renaming
```
huh, not sure what happened to the colon... will fix.

mmccool replied:
fixed

mmccool commented:
OK, some updates on points I think we have consensus on:
- Extended definition of Virtual Things to include interfacing to one or more other Things, etc.  See new defn for detail, but I'm trying to cover the use cases mentioned without making it TOO broad.  
- Added Connected Device, but this is a pure synonym for Device.   Bidirectionally cross-referenced.
- Updated Shadow to simplify the definition (although since I don't 100% agree we should remove the detail about that use case, I left it as a comment in the source, just in case we want to revert it or add it as a "for example").  Took out "advanced" when talking about Digital Twins.

mmccool commented:
So... in order to answer some questions in my head about Web Thing vs. Thing, I looked at the definition for "Thing or Web Thing" and have some concerns:
- It defines "virtual entity" as a "composition"
- It says "metadata and interfaces", that maybe should be "or" (at least)

I would prefer the very simple definition as follows:
- Thing: An entity described by a Thing Description.

I think the current definition of Thing Description is ok.  But if we just say "described by a Thing Description" then we don't have to repeat ourselves with "metadata and/or interfaces and/or links..." in the definition of Thing.  And since we say Virtual Things, Devices, and Services etc. are all a kind of Thing, the "virtual entity" and "physical entity" parts can be dropped.  Finally, "abstraction" is just confusing and I feel, unnecessary.

mmccool commented:
To clarify I'm not touching Web Thing.  We discussed this at length already and while I'm not convinced we need it, I'm ok with leaving it alone and that's the easiest option.

benfrancis commented:
Looks good to me, thanks.
